compactor/periodic.go

```diff
@@ -49,54 +49,64 @@ func NewPeriodic(h time.Duration, rg RevGetter, c Compactable) *Periodic {
 	return &Periodic{
 		clock:  clockwork.NewRealClock(),
 		period: h,
+		revs:   make([]int64, 0),
 		rg:     rg,
 		c:      c,
 	}
 }
 
-// periodDivisor divides Periodic.period in into checkCompactInterval duration
-const periodDivisor = 10
-
 func (t *Periodic) Run() {
 	t.ctx, t.cancel = context.WithCancel(context.Background())
-	t.revs = make([]int64, 0)
-	clock := t.clock
-	checkCompactInterval := t.period / time.Duration(periodDivisor)
+	interval := t.getInterval()
+
+	last := t.clock.Now()
 	go func() {
-		last := clock.Now()
 		for {
 			t.revs = append(t.revs, t.rg.Rev())
 			select {
 			case <-t.ctx.Done():
 				return
-			case <-clock.After(checkCompactInterval):
+			case <-t.clock.After(interval):
 				t.mu.Lock()
 				p := t.paused
 				t.mu.Unlock()
 				if p {
 					continue
 				}
 			}
-			if clock.Now().Sub(last) < t.period {
-				continue
-			}
-			rev, remaining := t.getRev()
-			if rev < 0 {
+
+			// wait up to initial given period
+			if t.clock.Now().Sub(last) < t.period {
 				continue
 			}
+			rev := t.revs[0]
+
 			plog.Noticef("Starting auto-compaction at revision %d (retention: %v)", rev, t.period)
 			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
-				t.revs = remaining
 				plog.Noticef("Finished auto-compaction at revision %d", rev)
+				// move to next sliding window
+				t.revs = t.revs[1:]
 			} else {
 				plog.Noticef("Failed auto-compaction at revision %d (%v)", rev, err)
-				plog.Noticef("Retry after %v", checkCompactInterval)
+				plog.Noticef("Retry after %v", interval)
```

Comment:
@yudai We are still discussing better ways to implement retry logic on compact failures. Any thoughts?

Reply:
Retrying in 1 hour sounds a bit longer to me.  5 minutes or `retension/10` would be more reasonable.

By the way, If In understand correctly, with this code, we don't slide `t.revs` in the error case, and `t.revs` gets a new value in the beginning of the next loop iteration  (line 65) anyway, meaning the `len(t.revs)` increases when an error happens. Eventually, the retention time is unexpectedly gets another 1 hour for each error. I think we should slide `t.revs` in the error case as well, if we keep this retry logic. Another idea is make line 82 like `rev := t.revs[len(t.revs) - interval]`.

The main point is that `t.revs = append(t.revs, t.rg.Rev())` at line 65 is expected to be executed every interval. To keep this promise, we cannot add another timed task for retry in the loop. 
We might use another gorotuine for retry, but I feel like it's an error prone way.
Maybe a better idea is that using a divided `interval` (by 10 or something, e.g. 6 minutes interval for 1hour retention period). And at line 82, we can do `rev := t.revs[len(t.revs) - interval*10]`? (we also need to modify line 89). In this case, we can retry in 1/10 retention time.

Reply:
I'm still not sure if my idea is feasible, I'll try implement it and see how it looks like.

Reply:
@yudai Sure. We will try implementing simplest retry logic first (for 3.3 backport). And iterate from there to improve. Thanks.

Reply:
The idea would look like this? However, I realized that if the retention time is small, the `subInterval` (`interval/10`) will be very very small. So this might not be a good idea to go.

https://gist.github.com/yudai/d976432a71e140598c08c91ae8569435

Reply:
Using two goroutine looks simpler in a sense and maybe most efficient.  The retry interval is `interval/10`.

https://gist.github.com/yudai/da0cf189c8c445973615671b33765c7d

@gyuho these two implementations are what I came up with. I personally like the way with two goroutines, but it's a big modification, so not sure if it's acceptable.

Reply:
> The retry interval is interval/10

I think this is a good idea since it won't be breaking change for 3.3.

And for two goroutines, let's do it once we backport this fix?
You can create a separate PR once we address this PR.
I will rework on this PR.

Thanks!

Reply:
Sure!